src/lib/comms/sol-http-server-impl-microhttpd.c

```diff
@@ -0,0 +1,587 @@
+    struct MHD_Response *r;
+    struct sol_http_param_value *value;
+
+    r = MHD_create_response_from_buffer(response->content.used, response->content.data,
+        MHD_RESPMEM_MUST_COPY);
+    if (!r)
+        return NULL;
+
+    SOL_HTTP_PARAM_FOREACH_IDX (&response->param, value, idx) {
+        int ret;
+        char buffer[512] = { 0 };
+
+        switch (value->type) {
+        case SOL_HTTP_PARAM_HEADER:
+            if (MHD_add_response_header(r, value->value.key_value.key, value->value.key_value.value) == MHD_NO) {
+                SOL_WRN("Could not add the header: %s", value->value.key_value.key);
+                goto err;
+            }
+            break;
+        case SOL_HTTP_PARAM_COOKIE:
+            ret = snprintf(buffer, sizeof(buffer), "%s=%s", value->value.key_value.key, value->value.key_value.value);
+            if (ret < 0 || ret >= (int)sizeof(buffer))
+                goto err;
+            if (MHD_add_response_header(r, MHD_HTTP_HEADER_SET_COOKIE, buffer) == MHD_NO) {
+                SOL_WRN("Could not add the cookie: %s", value->value.key_value.key);
+                goto err;
+            }
+            break;
+        case SOL_HTTP_PARAM_QUERY_PARAM:
+        case SOL_HTTP_PARAM_ALLOW_REDIR:
+        case SOL_HTTP_PARAM_TIMEOUT:
+        case SOL_HTTP_PARAM_AUTH_BASIC:
+        default:
+            break;
+        }
+    }
+
+    return r;
+
+err:
+    MHD_destroy_response(r);
+    return NULL;
+}
+
+static int
+post_iterator(void *data, enum MHD_ValueKind kind, const char *key,
+    const char *filename, const char *type,
+    const char *encoding, const char *value, uint64_t off, size_t size)
+{
+    struct sol_http_request *request = data;
+    char *v, *k;
+
+    if (!size)
+        return MHD_NO;
+
+    v = strdup(value);
+    SOL_NULL_CHECK(v, MHD_NO);
+
+    k = strdup(key);
+    SOL_NULL_CHECK_GOTO(k, err);
+
+    if (!sol_http_param_add(&request->params,
+        SOL_HTTP_REQUEST_PARAM_POST_FIELD(k, v))) {
+        SOL_WRN("Could not add %s key", key);
+        goto param_err;
+    }
+
+    return MHD_YES;
+
+param_err:
+    free(k);
+err:
+    free(v);
+    return MHD_NO;
+}
+
+static time_t
+process_if_modified_since(const char *value)
+{
+    char *s;
+    struct tm t;
+
+    s = strptime(value, "%a, %d %b %Y %H:%M:%S GMT", &t);
+    if (!s || *s != '\0')
+        return 0;
+
+    return timegm(&t);
+}
+
+static int
+headers_iterator(void *data, enum MHD_ValueKind kind, const char *key, const char *value)
+{
+    struct sol_http_request *request = data;
+    char *v, *k;
+
+    v = strdup(value);
+    SOL_NULL_CHECK(v, MHD_NO);
+
+    k = strdup(key);
+    SOL_NULL_CHECK_GOTO(k, err);
+
+    switch (kind) {
+    case MHD_HEADER_KIND:
+        if (!sol_http_param_add(&request->params,
+            SOL_HTTP_REQUEST_PARAM_HEADER(k, v)))
+            goto param_err;
+        if (streq(k, SOL_HTTP_PARAM_IF_SINCE_MODIFIED))
+            request->if_since_modified = process_if_modified_since(v);
```

Comment:
If `process_if_modified_since()` returns 0, nothing is done. Maybe `SOL_WRN()` and `goto param_err`?